required/latex-lab/latex-lab-math.dtx

```diff
@@ -2568,6 +2437,257 @@
+%      \cs{belowdisplayshortskip} to that they become negative and
+%      record that we have done so in the token list
+%      \cs{g_@@_skip_sign_tl} by setting it to \texttt{-}.
+%    \item
+%      But we only do that if both are non-negative. If either of them
+%      is negative we asume that this was deliberately done by the
+%      user to adjust the spacing around this particular display.
+%      Again  \cs{g_@@_skip_sign_tl} records that we haven't
+%      done the sign flipping by setting it to empty.
+%    \item
+%      Making the skips negative is essential, because the formula
+%      will be added using the special \cs{postdisplaypenalty} that
+%      doesn't allow a page break even though the real value (that we
+%      use later on) will probably do that. Thus the skip will add to
+%      the size of the display and not vanish into the margin and
+%      if it is positive \TeX{} might decide to put the whole formula
+%      onto the next page even though it might well fit.
+%    \item
+%      When we are passed the \verb=$$= \TeX{} has added, something
+%    like
+%\begin{verbatim}
+%   \penalty 100000        % our special value for \postdisplaypenalty
+%   \glue -10.0 plus -3pt  % the \belowdisplayskip (with sign flipped
+%\end{verbatim}
+%    \end{itemize}
+%      This means that at that point we can retrieve the skip with
+%    \cs{lastskip} and just have to flip the sign again to know how to
+%    correct it (no need to know if the normal or the short skip was
+%    added by \TeX{}) and the right penalty is available to us too as
+%    we have saved it away.
+%    
+%    
+%    
+% \begin{macro}{\@@_prepare_display_end:}
+%  
+%    Prepare to finish the formula and give control to \TeX{}. This is
+%    normally used directly in front of \verb=$$=  (\cs{eqno} or
+%    \cs{leqno}.
+%  
+% \changes{v0.6n}{2025/03/07}{Flipping the sign of the ``below skips'' at
+%    when the formula ends and not in \cs{everydisplay} (tagging/809)}
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_prepare_display_end: {  
+  \bool_lazy_or:nnTF
+    { \dim_compare_p:nNn \belowdisplayskip  < {0pt} }
+    { \dim_compare_p:nNn \belowdisplayshortskip  < {0pt} }
+%    \end{macrocode}
+%    Now sign flipping if one or both of the skips are already negative.  
+%    \begin{macrocode}
+    { \tl_gclear:N \g_@@_skip_sign_tl  }
+%    \end{macrocode}
+%    Otherwise flip the sign on both.
+%    \begin{macrocode}
+    {
+      \tl_gset:Nn \g_@@_skip_sign_tl {-}
+      \skip_set:Nn \belowdisplayskip      {-\belowdisplayskip}
+      \skip_set:Nn \belowdisplayshortskip {-\belowdisplayshortskip}
+    }
+%    \end{macrocode}
+%    For the skip it is enough to flip the sign, because we get the
+%    value back through \cs{lastskip} but for the
+%    \cs{postdisplaypenalty} change we have to  save the current value
+%    somewhere, because it may have been changed withing the display
+%    formula. This has to be done globally, because it is used after
```

Comment:
would it be safer to smuggle the value out with aftergroup or some such, thinking of (rare) cases of display math in a parbox in display math. Oh actually probably this is OK anyway as you are only saving the value from just before the end of the display to just after the `$$` so you can't have a nested display there.

Reply:
Yes, the setting happens only just before the formula ends so using a global setting is safe and simple.